pkg/operator/hive/configmap.go

```diff
@@ -0,0 +1,318 @@
+// cmInfo is a configMapInfo containing the string constants to be used for the configmap's name
+//   and the key therein that will correspond to the base filename when the configmap is mounted
+//   into the controller's container; and the data to be stored in the configmap.
+// namespacesToClean is a list of strings indicating former target namespaces from which this configmap
+//   is to be deleted.
+func (r *ReconcileHiveConfig) deployConfigMap(hLog log.FieldLogger, h resource.Helper, instance *hivev1.HiveConfig, cmInfo configMapInfo, namespacesToClean []string) (string, error) {
+	cmLog := hLog.WithField("configMap.name", cmInfo.name)
+
+	// Delete the configmap from previous target namespaces
+	for _, ns := range namespacesToClean {
+		cmLog.WithField("namespace", ns).Info("Deleting configmap from old target namespace")
+		// h.Delete already no-ops for IsNotFound
+		// TODO: Something better than hardcoding apiVersion and kind.
+		if err := h.Delete("v1", "ConfigMap", ns, cmInfo.name); err != nil {
+			return "", errors.Wrapf(err, "error deleting configmap/%s from old target namespace %s", cmInfo.name, ns)
+		}
+	}
+
+	cm := &corev1.ConfigMap{}
+	cm.Name = cmInfo.name
+	cm.Namespace = getHiveNamespace(instance)
+	cm.Data = make(map[string]string)
+
+	if cmInfo.setData != nil {
+		if err := cmInfo.setData(instance, &cm.Data); err != nil {
+			return "", err
+		}
+	} else {
+		data, err := cmInfo.getData(instance)
+		if err != nil {
+			return "", err
+		}
+		if reflect.TypeOf(data).Kind() != reflect.Ptr {
+			panic("Non-pointer value returned from configMapInfo.data(). This is a bug.")
+		}
+		if !reflect.ValueOf(data).IsNil() {
+			data, err := json.Marshal(data)
+			if err != nil {
+				return "", errors.Wrapf(err, "failed to marshal %s config", cmInfo.name)
+			}
+			cm.Data[cmInfo.nameKey] = string(data)
+		}
+	}
+
+	result, err := util.ApplyRuntimeObjectWithGC(h, cm, instance)
+	if err != nil {
+		cmLog.WithError(err).Error("error applying configmap")
+		return "", err
+	}
+	cmLog.WithField("result", result).Info("configmap applied")
+
+	cmLog.Info("Hashing configmap data onto a hive deployment annotation")
+
+	return computeHash(cm.Data), nil
+}
+
+func addConfigVolume(podSpec *corev1.PodSpec, cmInfo configMapInfo) {
+	volume := corev1.Volume{}
+	volume.Name = cmInfo.name
+	volume.ConfigMap = &corev1.ConfigMapVolumeSource{
+		LocalObjectReference: corev1.LocalObjectReference{
+			Name: cmInfo.name,
+		},
+	}
+	if cmInfo.volumeSourceOptional {
+		optional := true
+		volume.ConfigMap.Optional = &optional
+	}
+	volumeMount := corev1.VolumeMount{
+		Name:      cmInfo.name,
+		MountPath: cmInfo.mountPath,
+	}
+	envVar := corev1.EnvVar{
+		Name:  cmInfo.envVar,
+		Value: fmt.Sprintf("%s/%s", cmInfo.mountPath, cmInfo.nameKey),
+	}
+	podSpec.Volumes = append(podSpec.Volumes, volume)
+	podSpec.Containers[0].VolumeMounts = append(podSpec.Containers[0].VolumeMounts, volumeMount)
```

Comment:
YAGNI? (I.e. fix that if/when it becomes an issue?)

Reply:
I guess my concern is more about how if/when it breaks, it'll be in a subtle way. The hive controller will still come up with a valid Pod.spec, but some mysteriously-not-working-feature is now failing after some unrelated change ("I only added this fancy new profiler side-car container, why doesn't log uploading work now???").

Since we have e2e coverage I suppose we avoid this escaping, but someone (me? ;) ) might be cranky when they see a hard-coded `Containers[0]` was the culprit.

Reply:
Using the container name, I can see equally sneaky failures if that name is changed. I can defend against that to some extent, but not 100%. What about iterating over all the containers and adding the volume to all of them?

Reply:
We can error with "container with name XYZ not found", and fail the hive-controller Deployment to make it more clear about where things aren't as expected.

Reply:
Heck, we could build the "hive-controller" container (instead of working at the Pod-object-level), and add it to the pod.spec.containers[] list as a fully formed entity, but that seems like even more refactoring b/c of the shared volumes list...